FOCUS #149: CommitmentDiscountId spec changes and supporting content (#166)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/commitmentdiscountid.md

# Commitment Discount ID

> A commitment-based discount is a contractual commitment for an amount of usage or spend, for a specified term, in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.

Reviewer: ```suggestion
A commitment-based discount is a contractual commitment for a minimum amount of usage or spend throughout a specified term, in exchange for discounted unit pricing on that amount. The commitment may be based on quantities of resource units or monetary value, with various payment options and timeframes.
```

- Additional clarity on discount (can be many unit prices, and only for the amount committed)
- Removed reference to specific resources in case providers do not or stop offering commitments on those resources.

Reviewer 2: I would prefer if we didn't have to say 'minimum amount' to make it more general- might be true today cloud commitments but if we think about the broader context, I could see someone saying you had to average over a period of time (e.g. for people, average 40 hours/week of a consultants time over some period)

Reviewer 3: Rather than minimum, it’s spend committed for a time period (not minimum spend).  I liked the original language before minimum was added.

Author: Something like:

> A commitment-based discount is a contractual commitment for an amount of usage or spend committed for a time period, in exchange for discounted unit pricing on that amount. The commitment may be based on quantities of resource units or monetary value, with various payment options and timeframes.

?

Reviewer: Updated the suggestion to remove the "minimum", to mitigate marginal changes to the spec if providers add additional terms to their definition. 

There is an important distinction to emphasize when removing "minimum", that the `term` of the entire commitment is not necessarily the `term` of the committed usage or spend amount. Hopefully "throughout" makes that clear.

In @udam-f2 's scenario, if applied to compute, then there is no distinction for commit term and CBD term : average of $40/hr Compute over a 1-year period (trued-up at year end). Rather than the current commit for an absolute $40/hr Compute over a 1-year period (enforceable per hour).

Author: By the way, should it be `time frame` instead of `timeframe`?

> A commitment-based discount is a contractual commitment to use a certain amount of a specific type of usage for a fixed term in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.

Reviewer 2: @kk09v @flanakin We need a single definition here (that hopefully also aligns with the FinOps Foundation definition - or that we suggest back to the foundation). CommitmentType PR has a different blurb describing this same thing.

Karl, could we prioritize Commitment-discounts in the glossary first since it's actively being worked on and referenced in many PRs.

Reviewer 2: Actually, to reduce inter-dependencies, @flanakin could we settle on one definition and go with it and Karl can do a sweep and move things into the glossary as he's doing the other ones as well.

> A commitment-based discount is a contractual commitment to use a certain amount of a specific type of usage for a fixed term in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.

Reviewer 2: ```suggestion
A commitment-based discount is a contractual commitment for an amount of usage or spend, for a specified term in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.
```

> A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider.

Reviewer 4: Add scenarios:
```suggestion
A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings per commitment-based discount. 
```

> A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings per commitment-based discount.

Author: Updated.

> A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings per commitment-based discount. 

Reviewer 2: ```suggestion
A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings due to commitment-based discounts.
```

> A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings per commitment-based discount. 

Reviewer 4: Remove trailing space.
```suggestion
A Commitment Discount ID is the identifier assigned to a commitment-based discount by the provider. Commitment Discount ID is commonly used for scenarios like calculating utilization and savings per commitment-based discount.
```

> The CommitmentDiscountId column MUST be present in the billing data. This column must be of type String and MUST NOT contain null values when a charge is related to a commitment-based discount. When a charge is not associated with a commitment-based discount, the column MUST be null. CommitmentDiscountId MUST be unique within the provider.

Reviewer 5: > when a charge is related to a commitment-based discount

Maybe: 'when a record is related to a commit-based discount'.

Indeed, in case of a resource covered by commitment-based discount, the charge (BilledCost) would be 0.
Maybe it could be explictly mentioned: the CommitDiscountId value must not be null if the record relates to the purchase of a commitment-based discount or if the record relates to the usage of a resource covered by a commitment-based discount.

Reviewer 2: I'm fairly sure we've referred to the record / line-item / row as a 'charge' in other parts of the doc. Does a charge have to be a non-zero number? We need to update other places if we want to start referring to the line items as a record.

Examples are:
- ChargePeriod / ChargeType applies to all rows
- "Billing Currency is an identifier that represents the currency that a charge for resources and/or services was billed in."
- "If a provider does not support a sub account construct (only has a billing account) or does support a sub account construct, but the charge does not apply to a sub account, the SubAccountId column MUST be null."

Reviewer 3: I don’t think a charge needs to be non-zero for this. I would still want to see that charge.

Reviewer 5: Thanks for the clarification.
Indeed, I had not noticed that before a 'charge' was a synonym for a record / line-item / row.

> The CommitmentDiscountId column MUST be present in the billing data. This column must be of type String and MUST NOT contain null values when a charge is related to a commitment-based discount. When a charge is not associated with a commitment-based discount, the column MUST be null. CommitmentDiscountId MUST be unique within the provider.

Reviewer 5: Maybe it can be mentioned that CommitmentDiscountID works in pair with CommitmentDiscountName.

Reviewer: While that is logically true for human-reading, an ID still retains its value as a dimension independent of the name. Would also look into how other ID/name dimensions are presented with this change, and I think lack of reference to each other makes sense.

Author: This is in keeping with our existing id/name pairs such as BillingAccountId, ResourceId, and SubAccountId. We might need to update those items as well.

Reviewer 5: Yes indeed, nothing is mentioned for the id fields in FOCUS v0.5.
So this could be done later overall and in a consistent way.

Reviewer 2: Yes, this was discussed earlier on - and we are going to move metrics and dimensions under a single 'columns' section and then add a section where we link 'related' columns and attributes. This will be done at the end - as a final pass. Good call-out though.
